src/Examples/Sequence/Treap.lagda.md

```diff
@@ -0,0 +1,162 @@
+# Treap
+
+In this file, we implement and verify the [treap](https://en.wikipedia.org/wiki/Treap) data structure.
+
+<!--
+```agda
+{-# OPTIONS --prop --rewriting #-}
+```
+-->
+
+First, let's import some stuff:
+```agda
+module Examples.Sequence.Treap where
+
+open import Algebra.Cost
+
+costMonoid = ℚ-CostMonoid
+open CostMonoid costMonoid
+import Data.Rational as ℚ
+
+open import Calf costMonoid
+
+open import Data.Interval
+open import Data.List.Properties
+open import Data.Nat.Properties as NatProp
+open import Examples.Decalf.ProbabilisticChoice
+
+open import Calf.Data.Product
+open import Calf.Data.Nat as Nat using (zero; suc; >-nonZero) 
+open import Calf.Data.List
+
+
+open import Function using (_$_)
+
+open import Relation.Nullary
+open import Relation.Nullary.Negation using (contradiction)
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality as Eq using (_≡_; refl; _≢_; module ≡-Reasoning; ≢-sym)
+```
+
+Here is the definition of treap:
+```agda
+data ITreap (A : tp⁺) : val (list A) → Set where
+  leaf : ITreap A []
+  node : {l₁ l₂ : val (list A)}
+    (t₁ : ITreap A l₁) (a : val A) (t₂ : ITreap A l₂)
+    → ITreap A (l₁ ++ [ a ] ++ l₂)
+itreap : (A : tp⁺) → val (list A) → tp⁺
+itreap A l = meta⁺ (ITreap A l)
+```
+
+Here is the implementation of join:
+```agda
+++-singleton : {T : Set} → (x : T) → (l₁ l₂ : List T) → l₁ ++ x ∷ l₂ ≡ l₁ ++ [ x ] ++ l₂
+++-singleton x [] l₂ = refl
+++-singleton x (x₁ ∷ l₁) l₂ = Eq.cong (λ l → x₁ ∷ l) (++-singleton _ _ _)
+
+i-join-lemma : {T : Set} → (a₁ a a₂ : T) → (l₁₁ l₁₂ l₂₁ l₂₂ : List T) → ((l₁₁ ++ a₁ ∷ l₁₂) ++ a ∷ l₂₁) ++ a₂ ∷ l₂₂ ≡ (l₁₁ ++ [ a₁ ] ++ l₁₂) ++ [ a ] ++ l₂₁ ++ [ a₂ ] ++ l₂₂
+i-join-lemma a₁ a a₂ l₁₁ l₁₂ l₂₁ l₂₂ = 
+  let open ≡-Reasoning in
+  begin
+    ((l₁₁ ++ a₁ ∷ l₁₂) ++ a ∷ l₂₁) ++ a₂ ∷ l₂₂
+  ≡⟨ Eq.cong ( λ l' →  (l' ++  a ∷ l₂₁) ++ a₂ ∷ l₂₂) (++-singleton a₁ l₁₁ l₁₂) ⟩
+    ((l₁₁ ++ [ a₁ ] ++ l₁₂) ++ a ∷ l₂₁) ++ a₂ ∷ l₂₂ 
+  ≡⟨ Eq.cong (λ l' → l' ++ (a₂ ∷ l₂₂)) (++-singleton a _ l₂₁) ⟩ 
+    ((l₁₁ ++ [ a₁ ] ++ l₁₂) ++ [ a ]  ++ l₂₁) ++ (a₂ ∷ l₂₂ )
+  ≡⟨ ++-singleton _ _ _ ⟩
+    ((l₁₁ ++ [ a₁ ] ++ l₁₂) ++ ([ a ] ++ l₂₁)) ++ ([ a₂ ] ++ l₂₂)
```

Comment:
(if the previous comment is correct, I think this is just the `++-assoc`?)
```suggestion
```

Reply:
yeah